src/en/cmanhua/src/eu/kanade/tachiyomi/extension/en/cmanhua/CManhua.kt

```diff
@@ -0,0 +1,612 @@
+            return jsonElement.jsonPrimitive.contentOrNull
+                ?: throw Exception(errorMessageFromBody(trimmed))
+        }
+
+        val data = jsonElement["d"]?.jsonPrimitive?.contentOrNull
+            ?: jsonElement["data"]?.jsonPrimitive?.contentOrNull
+            ?: jsonElement["html"]?.jsonPrimitive?.contentOrNull
+            ?: throw Exception(errorMessageFromBody(trimmed))
+
+        return data
+    }
+
+    private fun errorMessageFromBody(body: String): String {
+        val code = body.trim().toIntOrNull()
+        if (code != null) {
+            return errorMessage(code)
+        }
+
+        val jsonObject = runCatching { json.parseToJsonElement(body) }.getOrNull() as? JsonObject
+        val error = jsonObject
+            ?.get("error")
+            ?.jsonPrimitive
+            ?.contentOrNull
+
+        return error ?: "Failed to load chapter pages."
+    }
+
+    private fun String.toAbsoluteUrl(): String {
+        if (isBlank()) {
+            return ""
+        }
+        return when {
+            startsWith("http") -> this
+            startsWith("//") -> "https:$this"
+            else -> baseUrl + this
+        }
+    }
+
+    private fun String.toStatus(): Int {
+        return when (lowercase(Locale.ROOT)) {
+            "on going", "ongoing" -> SManga.ONGOING
+            "completed" -> SManga.COMPLETED
+            else -> SManga.UNKNOWN
+        }
+    }
+
+    private fun parseChapterNumber(name: String): Float {
+        return CHAPTER_NUMBER_REGEX.find(name)?.groupValues?.getOrNull(1)?.toFloatOrNull() ?: -1f
+    }
+
+    private open class UriPartFilter(
+        displayName: String,
+        private val options: Array<Pair<String, String>>,
+    ) : Filter.Select<String>(displayName, options.map { it.first }.toTypedArray()) {
+        fun toUriPart(): String = options[state].second
+    }
+
+    private class StatusFilter : UriPartFilter("Status", STATUS_OPTIONS)
+
+    private class SortFilter : UriPartFilter("Order by", SORT_OPTIONS)
+
+    private class ChapterCountFilter : UriPartFilter("Min chapters", CHAPTER_OPTIONS)
+
+    private class GenderFilter : UriPartFilter("For", GENDER_OPTIONS)
+
+    private class GenreFilter(genres: List<Genre>) : Filter.Group<Genre>("Genres", genres)
+
+    private class Genre(name: String, val id: String) : Filter.CheckBox(name)
+
+    private fun errorMessage(code: Int): String {
+        return when (code) {
+            3 -> "You do not have enough coins to unlock this chapter."
+            -1 -> "Chapter does not exist."
+            -2 -> "Chapter token expired."
+            -3 -> "Invalid parameters."
+            -4 -> "Login required to access this chapter."
+            -5 -> "Account is banned or does not exist."
+            else -> "Failed to load chapter pages."
+        }
+    }
+
+    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.ROOT).apply {
+        timeZone = TimeZone.getTimeZone("UTC")
+    }
```

Comment:
This can also be written as (you don't need to change it):
```suggestion
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.ROOT)
```